src/sentry/integrations/github/integration.py

```diff
@@ -399,57 +404,86 @@ def setup(self) -> None:
 
-        if not request.GET.get("state"):
-            state = pipeline.signature
 
-            redirect_uri = absolute_uri(
-                reverse("sentry-extension-setup", kwargs={"provider_id": "github"})
-            )
-            return self.redirect(
-                f"{ghip.get_oauth_authorize_url()}?client_id={github_client_id}&state={state}&redirect_uri={redirect_uri}"
+class OAuthLoginView(PipelineView):
+    def dispatch(self, request: Request, pipeline) -> HttpResponseBase:
+        with record_event(IntegrationPipelineViewType.OAUTH_LOGIN).capture() as lifecycle:
+            self.determine_active_organization(request)
+            lifecycle.add_extra(
+                "organization_id",
+                self.active_organization.organization.id if self.active_organization else None,
             )
 
-        # At this point, we are past the GitHub "authorize" step
-        if request.GET.get("state") != pipeline.signature:
-            return error(request, self.active_organization, error_short="Invalid state")
-
-        # similar to OAuth2CallbackView.get_token_params
-        data = {
-            "code": request.GET.get("code"),
-            "client_id": github_client_id,
-            "client_secret": github_client_secret,
-        }
-
-        # similar to OAuth2CallbackView.exchange_token
-        req = safe_urlopen(url=ghip.get_oauth_access_token_url(), data=data)
-
-        try:
-            body = safe_urlread(req).decode("utf-8")
-            payload = dict(parse_qsl(body))
-        except Exception:
-            payload = {}
-
-        if "access_token" not in payload:
-            return error(request, self.active_organization, error_short="Missing access token")
-
-        authenticated_user_info = get_user_info(payload["access_token"])
-        if "login" not in authenticated_user_info:
-            return error(request, self.active_organization, error_short="Missing login info")
+            ghip = GitHubIdentityProvider()
+            github_client_id = ghip.get_oauth_client_id()
+            github_client_secret = ghip.get_oauth_client_secret()
+
+            installation_id = request.GET.get("installation_id")
+            if installation_id:
+                pipeline.bind_state("installation_id", installation_id)
+
+            if not request.GET.get("state"):
+                state = pipeline.signature
+
+                redirect_uri = absolute_uri(
+                    reverse("sentry-extension-setup", kwargs={"provider_id": "github"})
+                )
+                return self.redirect(
+                    f"{ghip.get_oauth_authorize_url()}?client_id={github_client_id}&state={state}&redirect_uri={redirect_uri}"
+                )
+
+            # At this point, we are past the GitHub "authorize" step
+            if request.GET.get("state") != pipeline.signature:
+                lifecycle.record_failure({"failure_reason": GitHubInstallationError.INVALID_STATE})
```

Comment:
What does invalid state mean? Like we're in step 3 without doing step 2 of the pipeline?

Reply:
i'm not exactly sure... i'm just reusing the definition from before